tests/digest.bats

```diff
@@ -3,40 +3,60 @@
 load helpers
 
 fromreftest() {
-  _prefetch $1
   run_buildah from --quiet --pull --signature-policy ${TESTSDIR}/policy.json $1
   cid=$output
+
+  # If image includes '_v2sN', verify that image is schema version N
+  expected_schemaversion=$(expr "$1" : '.*_v2s\([0-9]\)')
+  if [ -n "$expected_schemaversion" ]; then
+      actual_schemaversion=$(imgtype -expected-manifest-type '*' -show-manifest $1 | jq .schemaVersion)
+      expect_output --from="$actual_schemaversion" "$expected_schemaversion" \
+                    ".schemaversion of $1"
+  fi
+
+  # This is all we test: basically, that buildah doesn't crash when pushing
   pushdir=${TESTDIR}/fromreftest
   mkdir -p ${pushdir}/{1,2,3}
   run_buildah push --signature-policy ${TESTSDIR}/policy.json $1 dir:${pushdir}/1
   run_buildah commit --signature-policy ${TESTSDIR}/policy.json $cid new-image
   run_buildah push --signature-policy ${TESTSDIR}/policy.json new-image dir:${pushdir}/2
   run_buildah rmi new-image
   run_buildah commit --signature-policy ${TESTSDIR}/policy.json $cid dir:${pushdir}/3
+
   run_buildah rm $cid
   rm -fr ${pushdir}
 }
 
 @test "from-by-digest-s1" {
-  fromreftest k8s.gcr.io/pause@sha256:bbeaef1d40778579b7b86543fe03e1ec041428a50d21f7a7b25630e357ec9247
+  fromreftest quay.io/libpod/testdigest_v2s1@sha256:816563225d7baae4782653efc9410579341754fe32cbe20f7600b39fc37d8ec7
 }
 
 @test "from-by-digest-s1-a-discarded-layer" {
-  fromreftest libpod/whalesay@sha256:2413c2ffc29fb01d51c27a91b804079995d6037eed9e4b632249fce8c8708eb4
+  IMG=quay.io/libpod/testdigest_v2s1_with_dups@sha256:70d6c767101c907aa251c21fded459c0f1a481685eb764ca2f7a6162e24dd81a
+
+  fromreftest ${IMG}
+
+  # Verify that image meets our expectations (duplicate layers)
+  # Surprisingly, we do this after fromreftest, not before, because fromreftest
+  # has to pull the image for us.
+  dups=$(imgtype -expected-manifest-type '*' -show-manifest ${IMG} | jq .fsLayers|grep blobSum|sort|uniq -cd)
+  if [[ -z "$dups" ]]; then
+      die "Image ${IMG} does not have any duplicate layers (expected: one dup)"
+  fi
```

Comment:
This test broke when we moved to using `libpod/whalesay`, since the new schema 1 image didn't have `fsLayers` and `history` lists in which with both the first and second entries were the identical.

I think we want to do a `jq '.fsLayers[].blobSum' | head -n 2 | uniq -cd` and `jq '.history[].v1Compatibility' | head -n 2 | uniq -cd` to detect that the image looks right, and find or make an image that looks more like `docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b` to use in the test.

Reply:
I'm super confused. Big wall o' questions coming.

First: when you say "this test broke", do you mean "the test started failing", or "the test was no longer testing what it was supposed to test"? I'm going to assume the latter, and that's exactly why my PR includes image tests that try to confirm image properties. (If you meant the former, or if I've misunderstood, please help me out).

Second... from my reading of git history, it looks like 86fa0803 introduced the `whalesay` image check and the next commit, ae7d2f35 less than one second later, also had something to do with duplicates. So I was expecting buildah-push to fail when run against 86fa0803^ -- and it does. But it fails in exactly the same way with ae7d2f35!
```console
++ ./buildah from --quiet --pull docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b
+ cid=whalesay-working-container-5
+ ./buildah push docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b dir:/tmp/bp/1
Getting image source signatures
Copying blob sha256:1154ba695078d29ea6c4e1adb55c463959cd77509adf09710e2315827d66271a
 188.05 MiB / 188.05 MiB [==================================================] 2s
Copying blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd
ERRO[0002] error unmounting /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
error pushing image "docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b" to "dir:/tmp/bp/1": error copying layers and metadata: Error reading blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd: error creating overlay mount to /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
```
What I am trying to do is figure out what is special about this whalesay, then get a version of buildah that fails with it, then get a version that passes, then try to craft my own custom image and see if we can use that instead of whalesay. Can you help me understand the buildah failures above, and help me isolate working/nonworking buildah commits for me to try out? Thank you!

Reply:
OBTW I think we should keep my `quay.io/libpod/testdigest_v2s1_with_dups` because with buildah ae7d2f35 it causes an interesting new failure:
```console
#  ./buildah commit testdigest_v2s1_with_dups-working-container-7 new-image
error committing container "testdigest_v2s1_with_dups-working-container-7" to "new-image": error copying layer and metadata: error reading configuration to write to image "containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]docker.io/library/new-image:latest": internal error: history lists 1 non-empty layers, but we have 4 layers on disk
```
So the exercise now is to create yet another magic image that causes old-buildah to fail but new-buildah to pass.

Reply:
DISCREPANCY SOLVED! I was using the default `overlay`, which of course doesn't work with buildah. Switching to `vfs` yields the expected results.

To recap: 86fa080 and ae7d2f35 are a two-part commit that addressed the whalesay image problem: after ae7d2f35, the `digest.bats` whalesay test passes. _However_, even with vfs and even with this commit, my hand-crafted `v2s1_with_dups` image causes buildah to fail:
```console
# ./buildah --storage-driver vfs commit testdigest_v2s1_with_dups-working-container-3 new-image
error committing container "testdigest_v2s1_with_dups-working-container-3" to "new-image": error copying layer and metadata: error reading configuration to write to image "containers-storage:[vfs@/var/lib/containers/storage+/var/run/containers/storage:vfs.override_kernel_check=true]docker.io/library/new-image:latest": internal error: history lists 1 non-empty layers, but we have 4 layers on disk
```
Given all this, I humbly suggest that my `testdigest_v2s1_with_dups` image is a better test than whalesay. Can we get this merged, and add new test cases (and images) as we find more failure cases?

Reply:
> I'm super confused. Big wall o' questions coming.
> 
> First: when you say "this test broke", do you mean "the test started failing", or "the test was no longer testing what it was supposed to test"? I'm going to assume the latter, and that's exactly why my PR includes image tests that try to confirm image properties. (If you meant the former, or if I've misunderstood, please help me out).

The latter, as the new image being used no longer had the characteristic that triggered the code paths we cared about for the test.

> Second... from my reading of git history, it looks like [86fa080](https://github.com/containers/buildah/commit/86fa0803e8991b2c970c1c248d8ff350c39a50ac) introduced the `whalesay` image check and the next commit, [ae7d2f3](https://github.com/containers/buildah/commit/ae7d2f354732f40e113163e8b2a214601517bae2) less than one second later, also had something to do with duplicates. So I was expecting buildah-push to fail when run against [86fa080](https://github.com/containers/buildah/commit/86fa0803e8991b2c970c1c248d8ff350c39a50ac)^ -- and it does. But it fails in exactly the same way with [ae7d2f3](https://github.com/containers/buildah/commit/ae7d2f354732f40e113163e8b2a214601517bae2)!
> 
> ```
> ++ ./buildah from --quiet --pull docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b
> + cid=whalesay-working-container-5
> + ./buildah push docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b dir:/tmp/bp/1
> Getting image source signatures
> Copying blob sha256:1154ba695078d29ea6c4e1adb55c463959cd77509adf09710e2315827d66271a
>  188.05 MiB / 188.05 MiB [==================================================] 2s
> Copying blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd
> ERRO[0002] error unmounting /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
> error pushing image "docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b" to "dir:/tmp/bp/1": error copying layers and metadata: Error reading blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd: error creating overlay mount to /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
> ```
> 
> What I am trying to do is figure out what is special about this whalesay, then get a version of buildah that fails with it, then get a version that passes, then try to craft my own custom image and see if we can use that instead of whalesay. Can you help me understand the buildah failures above, and help me isolate working/nonworking buildah commits for me to try out? 

The manifest in the image is a v2s1 manifest, and in that manifest, the first entries in both the `fsLayers` and `history` lists are identical to the second entries in those lists, which can be confirmed by using `skopeo inspect --raw`.  The duplicate values are in consecutive entries.

While `quay.io/libpod/testdigest_v2s1_with_dups@sha256:70d6c767101c907aa251c21fded459c0f1a481685eb764ca2f7a6162e24dd81a` looks like it would make a useful test, particularly if we fail to commit a new image based on it (or used to fail, but don't any more), it doesn't trigger duplicate elimination logic in containers/image that logic in buildah needs to agree with.

This can be confirmed by counting the number of `fsLayer` entries in the v2s1 manifest using `skopeo inspect --raw`, and comparing that number to the number of `rootfs.diff_ids` values we get when we ask `skopeo inspect --config` to produce a v2s2 configuration blob, which it has to generate from the manifest.  Part of the process for generating that configuration blob includes dropping the duplicates from the manifest.

`skopeo` doesn't have a CLI option for requesting the manifest in v2s2 format, but then this is the first occasion I can think of when one might come in handy.

Reply:
Thanks for your response. If I understand you correctly, the magic characteristics that are desired are that `.fsLayers` be greater than `.rootfs.diff_ids`, and any image with those characteristics should make a test case. Is that correct? 

If so, is there a trick to getting such an image uploaded to quay.io? I've hand-crafted an image (by editing `manifest.json`) such that I believe it matches those constraints:
```console
# skopeo inspect --raw containers-storage:7aaf3 | jq '.fsLayers | length'
WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled
5

# skopeo inspect --config containers-storage:7aaf | jq '.rootfs.diff_ids | length'
WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled
4
```
...but when I push it to quay.io, and pull it back, `.fsLayers` is back to four. I've tried the following push commands:
```console
# skopeo copy containers-storage:localhost/testdigest_v2s1_with_dups:latest docker://quay.io/libpod/testdigest_v2s1_with_dups:$(date +%Y%m%d)

# podman push localhost/testdigest_v2s1_with_dups:latest quay.io/libpod/testdigest_v2s1_with_dups:$(date +%Y%m%d)
```
And I always pull with `buildah`.

Is this something `quay.io` is doing, or is it my mistake in podman/skopeo push? Is there some way to get this image uploaded to quay, or is this a futile exercise?

Reply:
> Thanks for your response. If I understand you correctly, the magic characteristics that are desired are that `.fsLayers` be greater than `.rootfs.diff_ids`, and any image with those characteristics should make a test case. Is that correct?

No, their lengths relative to each other are not a key factor.  It's that the first two elements of the `fsLayers` are identical to each other, and the first two elements of the `history` are identical to each other.  The `history` values even have the same `id` and `parent` values in their `v1Compatibility` fields.

> Is this something `quay.io` is doing, or is it my mistake in podman/skopeo push? Is there some way to get this image uploaded to quay, or is this a futile exercise?

It appears to be something that happens at the client before it attempts to upload the manifest, normally a good thing.

Reply:
Thanks, Nalin. Using a `skopeo` doctored with your gist, I've created and uploaded `quay.io/libpod/testdigest_v2s1_with_dups:20200323`; and I've confirmed that upon pull, 'buildah manifest' includes the duplicate layers. 

Could I ask you to pull it and confirm that it meets your requirements for a test image? If it does, I'll spend the rest of the time updating the test scripts and documentation then resubmit for review.

Reply:
Yup, looks good!  Inspecting with `skopeo inspect --raw` shows the duplicates, and they're omitted from the configuration it synthesizes when I use `skopeo inspect --config`.

Reply:
Does this mean we can merge?

Reply:
Not right now. I made the stupid, stupid, stupid mistake of rebasing against master. Tests are broken, and it's nothing to do with anything in this PR. Right now it looks like #2245 but I can't understand how that passed CI. I'll keep investigating.